File: javascript/jssearch/src/searchsettings.js
/*
 * searchsettings.js
 *
 * represents the settings to use when performing the search
 */

var FileTypes = require('./filetypes.js').FileTypes;

function SearchSettings() {
    "use strict";
    let self = this;
    self.archivesOnly = false;
    self.debug = false;
    self.excludeHidden = true;
    self.firstMatch = false;
    self.inArchiveExtensions = [];
    self.inArchiveFilePatterns = [];
    self.inDirPatterns = [];
    self.inExtensions = [];
    self.inFilePatterns = [];
    self.inFileTypes = [];
    self.inLinesAfterPatterns = [];
    self.inLinesBeforePatterns = [];
    self.linesAfter = 0;
    self.linesAfterToPatterns = [];
    self.linesAfterUntilPatterns = [];
    self.linesBefore = 0;
    self.listDirs = false;
    self.listFiles = false;
    self.listLines = false;
    self.maxLineLength = 150;
    self.multilineSearch = false;
    self.outArchiveExtensions = [];
    self.outArchiveFilePatterns = [];
    self.outDirPatterns = [];
    self.outExtensions = [];
    self.outFilePatterns = [];
    self.outFileTypes = [];
    self.outLinesAfterPatterns = [];
    self.outLinesBeforePatterns = [];
    self.printResults = false;
    self.printUsage = false;
    self.printVersion = false;
    self.recursive = true;
    self.searchArchives = false;
    self.searchPatterns = [];
    self.startPath = "";
    self.uniqueLines = false;
    self.verbose = false;

    const addExtensions = function (exts, arr) {
        let xs = exts;
        if (typeof(exts) === 'string') {
            xs = exts.split(/,/);
        }
        xs.filter(x => x !== '').forEach(x => arr.push(x));
    };
    self.addInExtension = function (ext) {
        addExtensions(ext, self.inExtensions);
    };
    self.addOutExtension = function (ext) {
        addExtensions(ext, self.outExtensions);
    };
    const addPatterns = function (patterns, arr) {
        if (typeof(patterns) === 'string') {
            arr.push(new RegExp(patterns));
        } else if (patterns.constructor === Array) {
            patterns.forEach(p => arr.push(new RegExp(p)));
        }
    };
    self.addInDirPattern = function (pattern) {
        addPatterns(pattern, self.inDirPatterns);
    };
    self.addOutDirPattern = function (pattern) {
        addPatterns(pattern, self.outDirPatterns);
    };
    self.addInFilePattern = function (pattern) {
        addPatterns(pattern, self.inFilePatterns);
    };
    self.addOutFilePattern = function (pattern) {
        addPatterns(pattern, self.outFilePatterns);
    };
    self.addSearchPattern = function (pattern) {
        addPatterns(pattern, self.searchPatterns);
    };
    self.addInArchiveExtension = function (ext) {
        addExtensions(ext, self.inArchiveExtensions);
    };
    self.addOutArchiveExtension = function (ext) {
        addExtensions(ext, self.outArchiveExtensions);
    };
    self.addInArchiveFilePattern = function (pattern) {
        addPatterns(pattern, self.inArchiveFilePatterns);
    };
    self.addOutArchiveFilePattern = function (pattern) {
        addPatterns(pattern, self.outArchiveFilePatterns);
    };
    self.addInLinesAfterPattern = function (pattern) {
        addPatterns(pattern, self.inLinesAfterPatterns);
    };
    self.addOutLinesAfterPattern = function (pattern) {
        addPatterns(pattern, self.outLinesAfterPatterns);
    };
    self.addInLinesBeforePattern = function (pattern) {
        addPatterns(pattern, self.inLinesBeforePatterns);
    };
    self.addOutLinesBeforePattern = function (pattern) {
        addPatterns(pattern, self.outLinesBeforePatterns);
    };
    self.addLinesAfterToPattern = function (pattern) {
        addPatterns(pattern, self.linesAfterToPatterns);
    };
    self.addLinesAfterUntilPattern = function (pattern) {
        addPatterns(pattern, self.linesAfterUntilPatterns);
    };

    const addFileTypes = function (filetypes, arr) {
        if (typeof(filetypes) === 'string') {
            filetypes.split(/,/).filter(ft => ft !== '').
                forEach(ft => arr.push(FileTypes.fromName(ft)));
        } else if (filetypes.constructor === Array) {
            filetypes.forEach(ft => arr.push(FileTypes.fromName(ft)));
        }
    };
    self.addInFileType = function (filetype) {
        addFileTypes(filetype, self.inFileTypes);
    };
    self.addOutFileType = function (filetype) {
        addFileTypes(filetype, self.outFileTypes);
    };

    self.setArchivesOnly = function () {
        self.setArchivesOnlyBool(true);
    };

    self.setArchivesOnlyBool = function (b) {
        self.archivesOnly = b;
        if (b) self.searchArchives = b;
    };

    self.setDebug = function () {
        self.setDebugBool(true);
    };

    self.setDebugBool = function (b) {
        self.debug = b;
        if (b) self.verbose = b;
    };

    const listToString = function (name, lst) {
        return `${name}=["${lst.join('","')}"]`;
    };

    self.toString = function () {
        return 'SearchSettings(' +
            'archivesOnly=' + self.archivesOnly +
            ', debug=' + self.debug +
            ', excludeHidden=' + self.excludeHidden +
            ', firstMatch=' + self.firstMatch +
            ', ' + listToString('inArchiveExtensions', self.inArchiveExtensions) +
            ', ' + listToString('inArchiveFilePatterns', self.inArchiveFilePatterns) +
            ', ' + listToString('inDirPatterns', self.inDirPatterns) +
            ', ' + listToString('inExtensions', self.inExtensions) +
            ', ' + listToString('inFilePatterns', self.inFilePatterns) +
            ', ' + listToString('inFileTypes', self.inFileTypes) +
            ', ' + listToString('inLinesAfterPatterns', self.inLinesAfterPatterns) +
            ', ' + listToString('inLinesBeforePatterns', self.inLinesBeforePatterns) +
            ', linesAfter=' + self.linesAfter +
            ', ' + listToString('linesAfterToPatterns', self.linesAfterToPatterns) +
            ', ' + listToString('linesAfterUntilPatterns', self.linesAfterUntilPatterns) +
            ', linesBefore=' + self.linesBefore +
            ', listDirs=' + self.listDirs +
            ', listFiles=' + self.listFiles +
            ', listLines=' + self.listLines +
            ', maxLineLength=' + self.maxLineLength +
            ', multilineSearch=' + self.multilineSearch +
            ', ' + listToString('outArchiveExtensions', self.outArchiveExtensions) +
            ', ' + listToString('outArchiveFilePatterns', self.outArchiveFilePatterns) +
            ', ' + listToString('outDirPatterns', self.outDirPatterns) +
            ', ' + listToString('outExtensions', self.outExtensions) +
            ', ' + listToString('outFilePatterns', self.outFilePatterns) +
            ', ' + listToString('outFileTypes', self.outFileTypes) +
            ', ' + listToString('outLinesAfterPatterns', self.outLinesAfterPatterns) +
            ', ' + listToString('outLinesBeforePatterns', self.outLinesBeforePatterns) +
            ', printResults=' + self.printResults +
            ', printVersion=' + self.printVersion +
            ', recursive=' + self.recursive +
            ', searchArchives=' + self.searchArchives +
            ', ' + listToString('searchPatterns', self.searchPatterns) +
            ',  startPath="' + self.startPath + '"' +
            ', uniqueLines=' + self.uniqueLines +
            ', verbose=' + self.verbose +
            ')';
    };
}

exports.SearchSettings = SearchSettings;

File: swift/swiftsearch/Sources/swiftsearch/SearchSettings.swift
//
//  SearchSettings.swift
//  swiftsearch
//
//  Created by Cary Clark on 5/12/15.
//  Copyright (c) 2015 Cary Clark. All rights reserved.
//

// import Foundation

public enum DefaultSettings {
    public static let archivesOnly = false
    public static let colorize = true
    public static let debug = false
    public static let excludeHidden = true
    public static let firstMatch = false
    public static let linesAfter = 0
    public static let linesBefore = 0
    public static let listDirs = false
    public static let listFiles = false
    public static let listLines = false
    public static let maxLineLength = 150
    public static let multiLineSearch = false
    public static let printResults = true
    public static let printUsage = false
    public static let printVersion = false
    public static let recursive = true
    public static let searchArchives = false
    public static let startPath: String? = ""
    public static let textFileEncoding: String = "UTF-8"
    public static let uniqueLines = false
    public static let verbose = false
}

public class SearchSettings: CustomStringConvertible {
    public var archivesOnly: Bool = DefaultSettings.archivesOnly
    public var colorize: Bool = DefaultSettings.colorize
    public var debug: Bool = DefaultSettings.debug
    public var excludeHidden: Bool = DefaultSettings.excludeHidden
    public var firstMatch: Bool = DefaultSettings.firstMatch
    public var linesAfter = DefaultSettings.linesAfter
    public var linesBefore = DefaultSettings.linesBefore
    public var listDirs: Bool = DefaultSettings.listDirs
    public var listFiles: Bool = DefaultSettings.listFiles
    public var listLines: Bool = DefaultSettings.listLines
    public var maxLineLength = DefaultSettings.maxLineLength
    public var multiLineSearch: Bool = DefaultSettings.multiLineSearch
    public var printResults: Bool = DefaultSettings.printResults
    public var printUsage: Bool = DefaultSettings.printUsage
    public var printVersion: Bool = DefaultSettings.printVersion
    public var recursive: Bool = DefaultSettings.recursive
    public var searchArchives: Bool = DefaultSettings.searchArchives
    public var startPath = DefaultSettings.startPath
    public var textFileEncoding = DefaultSettings.textFileEncoding
    public var uniqueLines: Bool = DefaultSettings.uniqueLines
    public var verbose: Bool = DefaultSettings.verbose

    public var inArchiveExtensions = Set<String>()
    public var inArchiveFilePatterns = [Regex]()
    public var inDirPatterns = [Regex]()
    public var inExtensions = Set<String>()
    public var inFilePatterns = [Regex]()
    public var inFileTypes = [FileType]()
    public var inLinesAfterPatterns = [Regex]()
    public var inLinesBeforePatterns = [Regex]()
    public var linesAfterToPatterns = [Regex]()
    public var linesAfterUntilPatterns = [Regex]()
    public var outArchiveExtensions = Set<String>()
    public var outArchiveFilePatterns = [Regex]()
    public var outDirPatterns = [Regex]()
    public var outExtensions = Set<String>()
    public var outFilePatterns = [Regex]()
    public var outFileTypes = [FileType]()
    public var outLinesAfterPatterns = [Regex]()
    public var outLinesBeforePatterns = [Regex]()
    public var searchPatterns = [Regex]()

    public init() {}

    fileprivate func splitExtensions(_ exts: String) -> [String] {
        exts.split { $0 == "," }.map { String($0) }
    }

    public func addInArchiveExtension(_ exts: String) {
        for ext in splitExtensions(exts) {
            inArchiveExtensions.insert(ext)
        }
    }

    public func addInExtension(_ exts: String) {
        for ext in splitExtensions(exts) {
            inExtensions.insert(ext)
        }
    }

    public func addOutArchiveExtension(_ exts: String) {
        for ext in splitExtensions(exts) {
            outArchiveExtensions.insert(ext)
        }
    }

    public func addOutExtension(_ exts: String) {
        for ext in splitExtensions(exts) {
            outExtensions.insert(ext)
        }
    }

    public func setArchivesOnly(_ archivesOnly: Bool) {
        self.archivesOnly = archivesOnly
        if archivesOnly { searchArchives = true }
    }

    public func setDebug(_ debug: Bool) {
        self.debug = debug
        if debug { verbose = true }
    }

    public var description: String {
        let s = "SearchSettings(" +
            "archivesOnly=\(archivesOnly)" +
            ", colorize=\(colorize)" +
            ", debug=\(debug)" +
            ", excludeHidden=\(excludeHidden)" +
            ", firstMatch=\(firstMatch)" +
            ", inArchiveExtensions=\(setToString(inArchiveExtensions))" +
            ", inArchiveExtensions=\(setToString(inArchiveExtensions))" +
            ", inDirPatterns=\(arrayToString(inDirPatterns))" +
            ", inExtensions=\(setToString(inExtensions))" +
            ", inFilePatterns=\(arrayToString(inFilePatterns))" +
            ", inFileTypes=\(arrayToString(inFileTypes.map { FileTypes.toName($0) }))" +
            ", inLinesAfterPatterns=\(arrayToString(inLinesAfterPatterns))" +
            ", inLinesBeforePatterns=\(arrayToString(inLinesBeforePatterns))" +
            ", linesAfterToPatterns=\(arrayToString(linesAfterToPatterns))" +
            ", linesAfterUntilPatterns=\(arrayToString(linesAfterUntilPatterns))" +
            ", linesAfter=\(linesAfter)" +
            ", linesBefore=\(linesBefore)" +
            ", listDirs=\(listDirs)" +
            ", listFiles=\(listFiles)" +
            ", listLines=\(listLines)" +
            ", maxLineLength=\(maxLineLength)" +
            ", outArchiveExtensions=\(setToString(outArchiveExtensions))" +
            ", outArchiveExtensions=\(setToString(outArchiveExtensions))" +
            ", outDirPatterns=\(arrayToString(outDirPatterns))" +
            ", outExtensions=\(setToString(outExtensions))" +
            ", outFilePatterns=\(arrayToString(outFilePatterns))" +
            ", outFileTypes=\(arrayToString(outFileTypes.map { FileTypes.toName($0) }))" +
            ", outLinesAfterPatterns=\(arrayToString(outLinesAfterPatterns))" +
            ", outLinesBeforePatterns=\(arrayToString(outLinesBeforePatterns))" +
            ", printResults=\(printResults)" +
            ", printUsage=\(printUsage)" +
            ", printVersion=\(printVersion)" +
            ", recursive=\(recursive)" +
            ", searchArchives=\(searchArchives)" +
            ", searchPatterns=\(arrayToString(searchPatterns))" +
            ", startPath=\"\(startPath!)\"" +
            ", textFileEncoding=\"\(textFileEncoding)\"" +
            ", uniqueLines=\(uniqueLines)" +
            ", verbose=\(verbose)" +
            ")"
        return s
    }
}
